modules/tinymce/src/core/main/ts/mode/Readonly.ts

```diff
@@ -38,23 +39,50 @@ const registerReadOnlyInputBlockers = (editor: Editor): void => {
     }
   });
 
-  editor.on('input', (e) => {
-    if (!e.isComposing && isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+  // Prevent undo level additions during composition
+  editor.on('input compositionend', (e) => {
+    if (isReadOnly(editor)) {
+      // Skip undo level addition during composition
+      if (!e.isComposing) {
+        const undoLevel = editor.undoManager.add();
+        if (Type.isNonNullable(undoLevel)) {
+          editor.undoManager.undo();
+        }
       }
     }
   });
 
-  editor.on('compositionend', () => {
+  // Set up mutation observer to detect and revert unintended changes
+  const observer = new MutationObserver((mutations) => {
     if (isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+      // Skip if we're in an undo transaction (annotation changes)
+      if (editor.undoManager.typing || editor.undoManager.data.length > 0) {
+        return;
       }
+
+      mutations.forEach((mutation) => {
+        if (mutation.type === 'characterData' || mutation.type === 'childList') {
+          // Revert the changes by restoring the previous content
+          const undoLevel = editor.undoManager.add();
+          if (Type.isNonNullable(undoLevel)) {
+            editor.undoManager.undo();
+          }
+        }
+      });
     }
   });
+
+  editor.on('init', () => {
+    observer.observe(editor.getBody(), {
```

Comment:
Do we still need this if we're blocking the composition events? Even with the readonly check this seems very expensive to have permanently attached to every editor.

If the answer is we do need it, maybe have it only attached during composition 🤔

Reply:
yeah we do need the observer to not show any inputed content while `composition`, I've updated this to start at `compositionstart` and disconnect at `compositionend`

Reply:
Composition events can't be cancelled. See spec: https://w3c.github.io/uievents/#event-type-compositionend

However we can tigger a undo on composition end that way we might not need the observer.

Also if you use observer then you should only observe if we are in a readonly state this code seems to always observe that and it could be slow.

Reply:
I added the change suggested.

There's some issues with moving undo to composition end, occasionally and frequently there will be some content be inserted into the text area using Japanese. So I kept the observer, and wrapped it in `isReadOnly()`